src/software/ai/navigator/path_planner/hrvo/hrvo_simulator.h

```diff
@@ -0,0 +1,113 @@
+
+#include "software/ai/navigator/path_planner/hrvo/hrvo_agent.h"
+#include "software/ai/navigator/path_planner/hrvo/lv_agent.h"
+#include "proto/tbots_software_msgs.pb.h"
+#include "proto/visualization.pb.h"
+#include "software/geom/vector.h"
+#include "software/world/world.h"
+#include "software/geom/algorithms/intersection.h"
+#include "software/geom/algorithms/contains.h"
+#include "software/geom/algorithms/nearest_neighbor_search.hpp"
+
+class HRVOSimulator {
+public:
+    /**
+     * Constructor
+     * @param time_step The time step between each step of the simulator
+     * @param robot_constants The robot constants to be used for all Agents representing a
+     * robot
+     * @param friendly_team_colour The colour of the friendly team
+     */
+    explicit HRVOSimulator(double time_step, const RobotConstants_t &robot_constants,
+                           const TeamColour friendly_team_colour);
+
+    /**
+     * Reset all agents to match the state of the given world.
+     * Friendly robots will use the Hybrid Reciprocal algorithm to traverse.
+     * Enemy robots will go directly towards their goal without trying to avoid any
+     * obstacles
+     *
+     * @param world The world which the simulation should be based upon
+     */
+    void updateWorld(const World &world);
+
+    /**
+     * Reset all friendly agents goal points to match the path of the given primitive set
+     *
+     * @param new_primitive_set
+     */
+    void updatePrimitiveSet(const TbotsProto::PrimitiveSet &new_primitive_set);
+
+    void doStep(Duration time_step);
+
+    /**
+     * Assumption: friendly team is the only robots running HRVO,
+     * only enemy robots have robot id offset
+     * @param robot_id
+     */
+    void visualize(unsigned int robot_id);
+
+private:
+
+    /**
+     *      Adds a new Hybrid Reciprocal Agent to the simulation based on Robot.
+     *
+     * @param robot    The robot which this agent should be based on
+     * @param type     Whether this robot is FRIENDLY or ENEMY
+     *
+     */
+    void configureHRVORobot(const Robot &robot);
+
+    /**
+     *      Adds a new Linear Velocity Agent to the simulation based on Robot.
+     *
+     * @param robot       	The robot which this agent should be based on
+     * @param destination 	Destination for this robot
+     * @param type 		Whether this robot is FRIENDLY or ENEMY
+    */
+    void configureLVRobot(const Robot &robot, const Vector &destination,
+                          TeamSide type);
+
+    // PrimitiveSet which includes the path which each friendly robot should take
+    TbotsProto::PrimitiveSet primitive_set;
+
+    // Latest World which the simulator has received
+    std::optional<World> world;
+
+    // The robot constants which all agents will use
+    RobotConstants_t robot_constants;
+
+    // The colour of the friendly team
+    const TeamColour friendly_team_colour;
+
+    // robot id to agent
+    std::map<unsigned int, std::shared_ptr<Agent>> robots;
```

Comment:
I dont think we need shared pointers anymore. Though we should be careful using just a regular Agent. In some cases we would have to make sure to get a reference from the map, or else it would copy the Agent and any changes will not affect the actual stored agent.